pkg/lsp/service.go

```diff
@@ -34,12 +39,67 @@ type TextDocumentSyncKind int
 
 const (
 	TDSKNone        TextDocumentSyncKind = 0
-	TDSKFull                             = 1
-	TDSKIncremental                      = 2
+	TDSKFull        TextDocumentSyncKind = 1
+	TDSKIncremental TextDocumentSyncKind = 2
 )
 
+type TextDocumentSyncOptions struct {
+	OpenClose         bool                 `json:"openClose,omitempty"`
+	Change            TextDocumentSyncKind `json:"change"`
+	WillSave          bool                 `json:"willSave,omitempty"`
+	WillSaveWaitUntil bool                 `json:"willSaveWaitUntil,omitempty"`
+	Save              *SaveOptions         `json:"save,omitempty"`
+}
+
+// TextDocumentSyncOptions holds either a TextDocumentSyncKind or
+// TextDocumentSyncOptions. The LSP API allows either to be specified
+// in the (ServerCapabilities).TextDocumentSync field.
+type TextDocumentSyncOptionsOrKind struct {
```

Comment:
maybe document that the older type (number) is deprecated, but kept for backwards compatibility.

Man all these union types in the spec are really annoying in Go.